### PR TITLE
AccessToken.matching_token_for returns wrong results

### PR DESCRIPTION
**Pull request by swilgosz**
**Description**

If there is access_token and application already saved in database with matching scopes,
access_token_for returns first token, completely ignoring ‘scopes’ argument.

This happens, because in ’Doorkeeper::OAuth::ScopeChecker’ the ‘match?’ method
compares token scopes against application scopes if application scopes are present.

**Expected behaviour:**

If there is no token with scopes requested in ‘mathing_token_for’, ‘nil should be returned,
so new token could be created.

This affects custom authentication methods based on scope checking
and it is potential security problem.

**Resolves** #957

### CODE REVIEW

**Comment by piotrj** [comment]
@swilgosz I don't know the code that well and would have to do a more thorough investigation, but isn't the current behavior correct? 

Here is my reasoning (version 1):
1. User tries to authorize an app with empty set of scopes
2. He already has an access token with scope "read"
3. This means that he does not have to authorize this app cause he already authorized the app for a set of scopes that is larger than what he asked for now.

My reasoning (version 2):
1. User tries to authorize an app with empty set of scopes.
2. He already has an access token with scope "read".
3. Since he didn't specify any scopes (the ones specified are empty) then we fallback to app's scopes.  Let's say app's scope is "read"
4. This means that the token is valid, the user already authorized the app and we can successfully return.

Nevertheless, I think there is a bug there. But a different one that you are describing. The problem is that if application has some default scopes, then we never take the scopes passed in params into considerations. I guess there was some misunderstanding between 417b9d57 and then some commit somewhere else. The `ScopeChecker.match?` method's second argument is called `server_scopes` but we pass the scopes we get in params there. Seems wrong. I need to dig deeper there. 

So to sum up -  I don't think we want exact match of scopes. Cause if you are asking for 'read' and you have already access token that has scopes 'read write' then this AccessToken matches the request scopes (and has more). 

Thoughts?

**Comment by toupeira** [comment]
@swilgosz I think my changes in https://github.com/doorkeeper-gem/doorkeeper/pull/907 might have resolved this issue as well, please check the new behaviour.

**Comment by nbulaj** [comment]
Hi @swilgosz. Can you check the `master` branch please?
